mindsdb/interfaces/agents/langchain_agent.py

```diff
@@ -439,33 +457,52 @@ def _invoke_agent_executor_with_prompt(agent_executor, prompt):
         return pred_df
 
     def stream_agent(self, df: pd.DataFrame, agent_executor: AgentExecutor, args: Dict) -> Iterable[Dict]:
-        '''
-        Streams completion chunks for an agent.
+        base_template = args.get('prompt_template', args['prompt_template'])
+        input_variables = re.findall(r"{{(.*?)}}", base_template)
 
-        Args:
-            df (pd.DataFrame): DataFrame to use as messages input
-            agent_executor (AgentExecutor): Executor to use for streaming agent
-            args (Dict): Args to pass to agent
+        prompts, _ = prepare_prompts(df, base_template, input_variables, args.get('user_column', USER_COLUMN))
 
-        Returns:
-            chunks (Iterable[Dict]): Completion chunks for agent
-        '''
-        # Prefer prediction time prompt template, if available.
-        base_template = args.get('prompt_template', args['prompt_template'])
-        input_variables = []
-        matches = list(re.finditer("{{(.*?)}}", base_template))
-
-        for m in matches:
-            input_variables.append(m[0].replace('{', '').replace('}', ''))
-
-        base_template = base_template.replace('{{', '{').replace('}}', '}')
-        first_row = df.iloc[0]
-        prompt_template = PromptTemplate(input_variables=input_variables, template=base_template)
-        kwargs = {}
-        for col in input_variables:
-            kwargs[col] = first_row[col] if first_row[col] is not None else ''  # add empty quote if data is missing
-        prompt = prompt_template.format(**kwargs)
-        if not prompt:
-            return
-        for chunk in agent_executor.stream(prompt, config={'callbacks': self._get_agent_callbacks(args)}):
-            yield chunk
+        callbacks, context_callback = prepare_callbacks(self, args)
+
+        yield 'data: {}\n\n'.format(json.dumps({"type": "start", "prompt": prompts[0]}))
```

Comment:
a lot of json.dumps in this function. 
maybe let it return dict and do json.dumps in function who receives it (_completion_event_generator)